ffs/grpc/service.go

```diff
@@ -82,115 +201,169 @@ func (s *Service) Show(ctx context.Context, req *pb.ShowRequest) (*pb.ShowReply,
+// Info returns an Api information.
+func (s *Service) Info(ctx context.Context, req *pb.InfoRequest) (*pb.InfoReply, error) {
 	i, err := s.getInstanceByToken(ctx)
 	if err != nil {
 		return nil, err
 	}
-	c, err := cid.Decode(req.GetCid())
+
+	info, err := i.Info(ctx)
 	if err != nil {
 		return nil, err
 	}
-	log.Infof("adding cid %s", c)
-	jid, err := i.PushConfig(c)
-	if err != nil {
-		return nil, err
+
+	reply := &pb.InfoReply{
+		Info: &pb.InstanceInfo{
+			ID: info.ID.String(),
+			DefaultCidConfig: &pb.DefaultCidConfig{
+				Hot: &pb.HotConfig{
+					Enabled:       info.DefaultCidConfig.Hot.Enabled,
+					AllowUnfreeze: info.DefaultCidConfig.Hot.AllowUnfreeze,
+					Ipfs: &pb.IpfsConfig{
+						AddTimeout: int64(info.DefaultCidConfig.Hot.Ipfs.AddTimeout),
+					},
+				},
+				Cold: &pb.ColdConfig{
+					Enabled: info.DefaultCidConfig.Cold.Enabled,
+					Filecoin: &pb.FilConfig{
+						RepFactor:      int64(info.DefaultCidConfig.Cold.Filecoin.RepFactor),
+						DealDuration:   info.DefaultCidConfig.Cold.Filecoin.DealDuration,
+						ExcludedMiners: info.DefaultCidConfig.Cold.Filecoin.ExcludedMiners,
+						CountryCodes:   info.DefaultCidConfig.Cold.Filecoin.CountryCodes,
+						Renew: &pb.FilRenew{
+							Enabled:   info.DefaultCidConfig.Cold.Filecoin.Renew.Enabled,
+							Threshold: int64(info.DefaultCidConfig.Cold.Filecoin.Renew.Threshold),
+						},
+					},
+				},
+			},
+			Wallet: &pb.WalletInfo{
+				Address: info.Wallet.Address,
+				Balance: info.Wallet.Balance,
+			},
+			Pins: make([]string, len(info.Pins)),
+		},
 	}
+	for i, p := range info.Pins {
+		reply.Info.Pins[i] = p.String()
+	}
+	return reply, nil
+}
 
-	ch, err := i.Watch(jid)
+// Watch calls API.Watch
+func (s *Service) Watch(req *pb.WatchRequest, srv pb.API_WatchServer) error {
+	i, err := s.getInstanceByToken(srv.Context())
 	if err != nil {
-		return nil, fmt.Errorf("watching add cid created job: %s", err)
+		return err
 	}
-	defer i.Unwatch(ch)
-	for job := range ch {
-		if job.Status == ffs.Success {
-			break
-		} else if job.Status == ffs.Cancelled || job.Status == ffs.Failed {
-			return nil, fmt.Errorf("error adding cid: %s", job.ErrCause)
-		}
+
+	jids := make([]ffs.JobID, len(req.Jids))
+	for i, jid := range req.Jids {
+		jids[i] = ffs.JobID(jid)
 	}
-	return &pb.AddCidReply{}, nil
-}
 
-func receiveFile(srv pb.API_AddFileServer, writer *io.PipeWriter) {
+	updates, err := i.Watch(jids...)
+	defer i.Unwatch(updates)
+
 	for {
-		req, err := srv.Recv()
-		if err == io.EOF {
-			_ = writer.Close()
-			break
-		} else if err != nil {
-			_ = writer.CloseWithError(err)
-			break
-		}
-		_, writeErr := writer.Write(req.GetChunk())
-		if writeErr != nil {
-			writer.CloseWithError(writeErr)
+		select {
+		case <-srv.Context().Done():
+			return nil
+		case job, ok := <-updates:
+			if !ok {
+				return nil // do we want to return an error here? nil will result as EOF on the client i think
```

Comment:
Under normal circumstances, this shouldn't happen since the client should control this with the `Unwatch` call.
I'd add maybe a warning/error here just to see if that is happening and maybe the consequence of a bug.
I'd keep the `return nil` and let the client decide what to do in that case.